### PR TITLE
Pin #217 

### PR DESCRIPTION
**Pull request by mzuenni**
Implemented pinning, but I have some notes:

-  I am not sure if it is a good thing to add `cores` to so many functions...
- right now only submissions in fuzz/run are pinned but not during solution generation
- For interactive problems, all programs are pinned to the same core

### CODE REVIEW

**Comment by RagnarGrootKoerkamp** [comment]
Do you observe differences in timings? Ie more stable and/or faster?

File: bin/parallel.py [hunk 1]
```diff
@@ -31,15 +33,27 @@ def __init__(self, f, num_threads=True):
         self.finish = False
 
         if self.num_threads:
+            if pin:
+                # only use available cores
+                cores = list(os.sched_getaffinity(0))
```
Comment (RagnarGrootKoerkamp) [inline bin/parallel.py:38]: likely this only works on linuxy systems

File: bin/parallel.py [hunk 2]
```diff
@@ -31,15 +33,27 @@ def __init__(self, f, num_threads=True):
...
+                if self.num_threads > len(cores):
+                    self.num_threads = len(cores)
```
Comment (RagnarGrootKoerkamp) [inline bin/parallel.py:40]: do 1 less to give some space to the OS. could even pin bapctools itself to that thread, but maybe not needed.

File: bin/parallel.py [hunk 3]
```diff
@@ -31,15 +33,27 @@ def __init__(self, f, num_threads=True):
...
+
+                # sort cores by reversed id. If num_threads << len(cores) this ensures that we 
+                # use different physical cores instead of hyperthreads
+                num_bits = max(map(lambda c: c.bit_length(), cores))
+                cores.sort(key=lambda c: util.bit_reverse(c, num_bits))
```
Comment (RagnarGrootKoerkamp) [inline bin/parallel.py:45]: so on my 6-core 12-thread machine, i think 1 and 7 are on the same core, and 1..=6 are on distinct cores. 
I also always assumed hyperthreads would be consecutive but apparently not. (You can confirm this using `cat /proc/cpuinfo` or so i think.)

**Review by RagnarGrootKoerkamp — COMMENTED** [review]
Overall the code looks ok i suppose. as you say, it's mostly passing data through a lot of layers which ideed does not look so pretty :/

File: bin/parallel.py (hunk 3 above)
Comment (mzuenni) [reply bin/parallel.py]: hm you seem to be right...

**Comment by RagnarGrootKoerkamp** [comment]
Not sure exactly is 'thread local global variables' are a thing in python, but maybe you use those to pass the `cores` info?

Actually, you could possibly just pin each thread to a core the moment it's spawned. At least for programs the 'affinity' is inherited I think, and this may also work for threads? That would for sure simplify things, and running the python part of the threads on those cores as well doesn't seem like a drawback? (the python anyway is idle while the programs run. Apart from interactive problems)

**Comment by mzuenni** [comment]
As I understand the doc `setaffinity` is set for a `pid` and threads run under the same pid... But it still seems to work as far as I can see o.O

**Comment by RagnarGrootKoerkamp** [comment]
Hmm, `pid = 0` may work?

**Comment by mzuenni** [comment]
yes, that is what I tried and what seems to work... but I did not expect it to work... anyway I prefer this and will change it

**Comment by mzuenni** [comment]
I kept the code in the `limit_setter` even though we don`t use it right now...

**Comment by RagnarGrootKoerkamp** [comment]
Ok LGTM and works on my machine. Didn't see a speed difference on the one small problem i tested though. Maybe we should bump the default parellelism (`config.args.default_jobs`) from `os.cpu_count()/2` to `os.cpu_count()-1`? Or even better make it the number of physical cores instead of hyperthreads (if we can easily get that without extra libraries).

**Comment by mzuenni** [comment]
That doesn't seem to be easy... the only way I found so far is `psutil`.

**Comment by RagnarGrootKoerkamp** [comment]
yeah... nvm then. All looks good, but I'm still not 100% convinced we actually need this. Can you make some test to show that timings become faster/more stable? otherwise there's not much point

**Comment by mzuenni** [comment]
It speeds up java and interactive problems by ~50%

**Comment by RagnarGrootKoerkamp** [comment]
Oh that's pretty good! Ready to merge?

**Comment by mzuenni** [comment]
Yes, should be fine :)

**Comment by RagnarGrootKoerkamp** [comment]
Oh it needs rebasing

**Comment by mzuenni** [comment]
are you sure? ^^'

**Comment by RagnarGrootKoerkamp** [comment]
> This branch cannot be rebased due to conflicts

**Pull request merged**